lib/configurationSchema.js

```diff
@@ -0,0 +1,438 @@
+                type: 'string',
+              },
+              {
+                type: 'object',
+                additionalProperties: {
+                  type: 'string',
+                },
+              },
+            ],
+          },
+        ],
+      },
+    },
+    sortImports: {
+      default: true,
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'boolean',
+        },
+      ],
+    },
+    stripFileExtensions: {
+      default: ['.js', '.jsx', '.ts', '.tsx'],
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'array',
+          items: {
+            type: 'string',
+          },
+        },
+      ],
+    },
+    tab: {
+      default: '  ',
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'string',
+        },
+      ],
+    },
+    useRelativePaths: {
+      default: true,
+      anyOf: [
+        {
+          instanceof: 'Function',
+        },
+        {
+          type: 'boolean',
+        },
+      ],
+    },
+  },
+};
+
+function findGlobalsFromEnvironments(environments) {
+  const result = Object.keys(globals.builtin);
+
+  environments.forEach((environment) => {
+    const envGlobals = globals[environment];
+    if (!envGlobals) {
+      return;
+    }
+    result.push(...Object.keys(envGlobals));
+  });
+  return result;
+}
+
+export function getDefaultConfig() {
+  return Object.entries(SCHEMA.properties).reduce((acc, [k, v]) => {
+    if (typeof v !== `object` || v === null) {
+      throw new Error(`Expected schema key '${k}' to be an object`);
+    }
+    if (v.hasOwnProperty('default')) {
+      acc[k] = v.default;
+    }
+    return acc;
+  }, {});
+}
+
+export function validate(data) {
+  const ajv = new Ajv();
+  ajvInstanceof(ajv);
+
+  const validate = ajv.compile(SCHEMA);
+
+  if (!validate(data)) {
+    return {
+      error: true,
+      messages: validate.errors
+        .map((err) => {
+          // report unknown identifiers the same as we have done in the past
+          // so we can showcase that the move to Ajv didn't change previous
+          // behaviors
+          if (err.message === 'must NOT have additional properties') {
+            // remove the extraneous identifier so we avoid later errors in code
+            // instancePath is empty if root key
+            const rootKey =
+              err.instancePath.split('/')[1] || err.params.additionalProperty;
+            delete data[rootKey];
+
+            return (
+              'Unknown configuration: `' +
+              (err.instancePath ? err.instancePath + `.` : ``) +
+              err.params.additionalProperty +
+              '`'
+            );
+          } else {
+            // remove the failing identifier so we avoid later errors in code
+            const rootKey = err.instancePath.split('/')[1];
+            delete data[rootKey];
+
+            console.log(err);
```

Comment:
Should this use console.error?

Reply:
Ah, I left that in there by mistake. I've removed it now.

Logging things in low-level modules makes the tests very chatty, and I think that's anyway not the right place to communicate this information. We already return the error messages, so it is the responsibility of the caller function to do so, if needed.